bot/seasons/evergreen/nationday.py

```diff
@@ -0,0 +1,149 @@
+import datetime
+import json
+import logging
+from pathlib import Path
+from typing import List, Tuple
+
+import discord
+import requests
+from discord.ext.commands import Bot, Cog, Context, command
+
+from bot.pagination import ImagePaginator
+
+
+logger = logging.getLogger(__name__)
+# RESTful API to get countries info
+url = "https://restcountries.eu/rest/v2/alpha/"
+# white flag image to show during error
+white_flag = "https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/\
+White_flag_waving.svg/158px-White_flag_waving.svg.png"
+# url to get images of flags
+flag_url = "https://www.countryflags.io/"
+
+
+class NationDay(Cog):
+    """NationDay Cog that contains nationday command."""
+
+    def __init__(self, bot: Bot):
+        self.bot = Bot
+
+    with open(Path("bot/resources/evergreen/nationday/countries_by_day.json"), "r") as file:
+        by_day = json.load(file)
+
+    with open(Path("bot/resources/evergreen/nationday/iso_codes.json"), "r") as file:
+        iso_codes = json.load(file)
+
+    async def get_key(self, val: str) -> str:
+        """Get date for particular country, if present."""
+        for key, value in self.by_day.items():
+            if val in value:
+                return key
+        return None
+
+    async def get_specific_country(self, country: str) -> Tuple[List[Tuple[str, str]], str]:
+        """
+        Get Indepedence Day of given country.
+
+        Return Indepedence Day and page of information on country.
+        """
+        # Get Indepedence Day of specified country
+        date = await self.get_key(country)
+        if date:
+            page, img = await self.get_country_info(country)
+            return [(page, img)], date
+        return [(None, None)], None
+
+    async def country_today(self) -> List[Tuple[str, str]]:
+        """
+        Get current day [Month & Day].
+
+        Return pages of info and flags of countries.
+        """
+        # Get current date [Month and day]
+        date = datetime.datetime.today()
+        month = date.strftime("%B")
+        day = date.day
+        today_date = f'{month} {day}'
+        today_date = today_date.lower()
+        try:
+            # Get list of countries
+            countries = self.by_day[today_date]
+            countries = list(countries.split(','))
+        except Exception as ke:
```

Comment:
If you're expecting a `KeyError`, then only catch that one specific exception.

The reasoning for this is that you may have solvable bug in the lines above that will be silenced by the broad except and you won't know about it. For example, if you accidentally misspelled the `split` function and got an `AttributeError` there, the broad except will also silence that error, despite it being something that can easily fix.

Alternatively, you can safely get a value out of a dictionary like so:
```python
countries = self.by_day.get(today_date)

if countries is None:
    # Handle the case where the dict doesn't contain our key
    return

# This code is only reached if the key is present
```
This is just an alternative approach to consider - there is nothing inherently wrong about catching the `KeyError`, if done properly.

Reply:
So this should be alright, I suppose?
`try:
            # Get list of countries
            countries = self.by_day[today_date]
            countries = list(countries.split(','))
        except KeyError as ke:
            err_msg = f"**No countries have their independence day today.**"
            logger.warning(err_msg)
            return [(err_msg, white_flag)]
`

Reply:
That's a lot better, as it will no longer silence errors other than `KeyError`.

Since only the first line in the `try` block, the `countries = self.by_day[today_date]`, may raise the `KeyError`, the line after (`countries.split(...)`), could be moved out - below the `except` block.

This will help you communicate the intent of the code a lot better. To generalize:

```python
try:
    # A dangerous piece of code that may raise specifically KeyError
except KeyError as ke:
    # Handle error
    return

# Safe code that is not expected to raise anything,
# but should only run if the `try` block succeeds
```
It's a lot easier to read and reason about code written like this, because it's easier to see *exactly what* may raise the error.

Reply:
Yep, gotcha. I really appreciate the pointers!